File: uva_116.cpp
# include <iostream>
# include <climits>
#define M 15
#define N 105
using namespace std;
int main(void){
    int m, n;
    while(cin >> m >> n){
        long int graph[M+1][N+1]={0};
        int path[M+1][N+1]={0};
        
        for (int i=0; i<m; ++i){
            for (int j=0; j<n; ++j){
                cin >> graph[i][j];
            }
        }
        long int dp[M+1][N+1]={0};
        for (int col=n-1; col>=0; --col){
            for (int row=0; row<m; ++row){
                int up=(row-1+m)%m;
                int mid=row;
                int down=(row+1)%m;
                int min;
                if(dp[up][col+1]<dp[mid][col+1])
                    min=up;
                else if (dp[up][col+1]==dp[mid][col+1])
                    min=up<mid?up:mid;
                else 
                    min=mid;
                if(dp[min][col+1]>dp[down][col+1])
                    min=down;
                else if(dp[min][col+1]==dp[down][col+1])
                    min=down<min?down:min;
                path[row][col]=min;
                dp[row][col]=dp[min][col+1]+graph[row][col];
            }
        }
        int ans=0;
        for (int i=0; i<m; ++i){
            if(dp[i][0]<dp[ans][0])
                ans=i;
        }
        int output = dp[ans][0];
        for (int i = 0; i < n ; ++i){
            cout << ans+1;
            if (i!=n-1)
                cout << ' ';
            ans = path[ans][i];
        }
        cout << endl;
        cout << output << endl;
    }
    return 0;
}

File: uva_116.py
M=15
N=105
while True:
    graph=[[0] *105 for _ in range(M)]
    path=[[0] *105 for _ in range(M)]
    try:
        m, n = input().split()
        m, n=int(m), int(n)
    except:
        break
    row, col = 0, 0
    while True:
        data=input().split()
        for item in data:
            graph[row][col]=int(item)
            col+=1
            if col==n:
                col=0
                row+=1
        if row==m:
            break
    '''
    for i in range(m):
        for j in range(n):
            print("{} ".format(graph[i][j]), end='')
        print('')
    print('')
    '''
    dp=[[0] *105 for _ in range(M)]
    for col in range(n-1, -1, -1):
        for row in range(m):
            up=(row-1+m)%m
            mid=row
            down=(row+1)%m
            if dp[up][col+1]<dp[mid][col+1]:
                MIN=up
            elif dp[up][col+1]==dp[mid][col+1]:
                MIN=up if up<mid else mid
            else:
                MIN=mid
            if dp[MIN][col+1]>dp[down][col+1]:
                MIN=down
            elif dp[MIN][col+1]==dp[down][col+1]:
                MIN=down if down<MIN else MIN
            path[row][col]=MIN
            dp[row][col]=dp[MIN][col+1]+graph[row][col]
    ans=0
    for i in range(m):
        if dp[i][0]<dp[ans][0]:
            ans=i
    output=dp[ans][0]
    for i in range(n):
        print("{}".format(ans+1), end='')
        if i != n-1:
            print(' ', end='')
        ans=path[ans][i]
    print('')
    print(output)
